File: android_app/app/src/main/java/com/match/android/system/log/CachingLog.kt
package com.match.android.system.log

import com.match.android.system.log.LogLevel.D
import com.match.android.system.log.LogLevel.E
import com.match.android.system.log.LogLevel.I
import com.match.android.system.log.LogLevel.V
import com.match.android.system.log.LogLevel.W
import com.match.android.ui.common.LimitedSizeQueue
import kotlinx.coroutines.channels.ConflatedBroadcastChannel
import kotlinx.coroutines.channels.sendBlocking
import java.util.Date

class CachingLog : Log {
    val logs = ConflatedBroadcastChannel<LimitedSizeQueue<LogMessage>>(
        LimitedSizeQueue(1000)
    )

    override fun setup() {}

    override fun v(message: String, tag: LogTag?) {
        log(LogMessage(V, addTag(tag, message)))
    }

    override fun d(message: String, tag: LogTag?) {
        log(LogMessage(D, addTag(tag, message)))
    }

    override fun i(message: String, tag: LogTag?) {
        log(LogMessage(I, addTag(tag, message)))
    }

    override fun w(message: String, tag: LogTag?) {
        log(LogMessage(W, addTag(tag, message)))
    }

    override fun e(message: String, tag: LogTag?) {
        log(LogMessage(E, addTag(tag, message)))
    }

    private fun log(message: LogMessage) {
        logs.sendBlocking(logs.value.apply { add(message) })
    }

    private fun addTag(tag: LogTag?, message: String) =
        (tag?.let { "$it - " } ?: "") + message
}

data class LogMessage(val level: LogLevel, val text: String, val time: Date = Date())

enum class LogLevel(val text: String) {
    V("Verbose"),
    D("Debug"),
    I("Info"),
    W("Warn"),
    E("Error")
}

File: ios_app/ios_app/logs/CachingLog.swift
import Foundation
import Combine

struct LimitedSizeQueue<T> {
    public private(set) var array: [T] = []

    private let maxSize: Int

    init(maxSize: Int) {
        self.maxSize = maxSize
    }

    public mutating func add(value: T) {
        array.append(value)

        if array.count > maxSize {
            array.removeSubrange(0..<(array.count - maxSize - 1))
        }
    }

    public var isEmpty: Bool { return array.isEmpty }
}

class CachingLog: LogNonVariadicTags {
    let status = PassthroughSubject<String, Never>()

    let logs: CurrentValueSubject<LimitedSizeQueue<LogMessage>, Never> =
        CurrentValueSubject(LimitedSizeQueue<LogMessage>(maxSize: 1000))

    private let loggerSerialQueue = DispatchQueue(label: "com.match.logger")

   
    func setup() {}

    func v(_ message: String, _ tags: [LogTag]) {
        log(LogMessage(level: .v, text: addTag(tags: tags, message: message)))
    }

    func d(_ message: String, _ tags: [LogTag]) {
        log(LogMessage(level: .d, text: addTag(tags: tags, message: message)))
    }

    func i(_ message: String, _ tags: [LogTag]) {
        log(LogMessage(level: .i, text: addTag(tags: tags, message: message)))
    }

    func w(_ message: String, _ tags: [LogTag]) {
        log(LogMessage(level: .w, text: addTag(tags: tags, message: message)))
    }

    func e(_ message: String, _ tags: [LogTag]) {
        log(LogMessage(level: .e, text: addTag(tags: tags, message: message)))
    }

    private func log(_ message: LogMessage) {
        var queue = logs.value
        queue.add(value: message)
        logs.send(queue)
    }

    private func addTag(tags: [LogTag], message: String) -> String {
        let tagsStr = tags.map { "[\($0)]"}.joined(separator: " ")
        let tagPart = tagsStr.isEmpty ? "" : tagsStr + " "
        return tagPart + message
    }
}

struct LogMessage {
    let level: LogLevel
    let text: String
    let time: Date = Date()
}
